### PR TITLE
New CoTeacher Agentic Workflow

### PR DESCRIPTION
**Pull request by AaronSosaRamos**
## Description

Implemented a new Assistant Architecture integrating Google Gemini 2.0 experimental features with a CoTeacher implementation. This includes enhanced tool functionality and updates to improve usability. Additionally, documentation and unit tests were updated to reflect these changes.

## Related Issue

N/A

## Type of Change

Please select the type(s) of change that apply and delete those that do not.

- [ ] **Bug fix**: A non-breaking change that fixes an issue.
- [x] **New feature**: A non-breaking change that adds functionality.
- [ ] **Breaking change**: A change that causes existing functionality to not work as expected.
- [ ] **Documentation update**: Changes or updates to documentation.
- [ ] **Code style update**: Changes that do not affect the meaning of the code (e.g., formatting).
- [x] **Refactoring**: A code change that neither fixes a bug nor adds a feature.
- [ ] **Performance improvement**: A change that improves performance.
- [x] **Test enhancement**: Adding or updating tests; no production code change.
- [ ] **Chore**: Changes to the build process or auxiliary tools; no production code change.
- [ ] **Other**: (please describe)

### Key Changes:

1. **Assistant Architecture Overhaul**:  
   - Integrated **Google Gemini 2.0 exp flash** for advanced processing capabilities.

2. **CoTeacher Implementation**:  
   - Introduced a new **CoTeacher module** for collaborative learning and task-sharing workflows.
   - Enhanced communication efficiency between the assistant and users in educational environments.

3. **Documentation and Testing Updates**:  
   - Added comprehensive unit tests for the new workflows to ensure functionality and reliability.
   - Revised documentation for clarity and usability.

## How to Test

1. **Assistant Architecture Testing**:  
   - Run workflows utilizing the updated CoTeacher module.
   - Verify outputs for accuracy and coherence.

2. **CoTeacher Testing**:  
   - Test task-sharing scenarios and collaboration workflows.
   - Validate feedback loops between the CoTeacher and users.

### Expected Outcomes:
- Improved task accuracy across workflows.
- Seamless integration of CoTeacher with the assistant.

Request Sample:
```json
{
   "assistant_inputs":{
      "assistant_group":"classroom_support",
      "assistant_name":"co_teacher",
      "user_info":{
         "user_name":"Aaron Sosa",
         "user_age":26,
         "user_preference":"Senior AI Engineer"
      },
      "messages":[
         {
            "role":"human",
            "type":"text",
            "timestamp":"string",
            "payload":{
               "text":"Give me insights for personalized teaching"
            }
         },
         {
            "role":"ai",
            "type":"text",
            "timestamp":null,
            "payload":{
               "text":"Okay, Aaron, as a Senior AI Engineer, you're likely someone who values data-driven strategies and innovative solutions. Let's tailor some insights for personalized teaching with that in mind. Here's how you can approach this, leveraging your analytical and problem-solving skills:\n\n**1. Data-Driven Student Understanding:**\n\n*   **Analytics Approach:** Just like in AI projects, you can analyze student performance data to identify patterns and tailor your teaching methods. Look at areas where students excel and where they struggle. Use this data to understand the varying needs of your students.\n*   **Actionable Steps:**\n    *   Use formative assessments (quizzes, exit tickets, in-class activities) to collect data.\n    *   Create a spreadsheet or use a digital tool to track student performance across different topics or skills.\n    *   Identify individual student strengths, areas of improvement, and preferred learning styles.\n\n**2. Adaptive Content and Delivery:**\n\n*   **Modular Content Creation:** Think of your teaching material like modular code, where you can easily modify and adapt it. Develop a core set of learning materials and then adapt to student needs by offering multiple versions (different problem sets, examples, or explanations).\n*   **Actionable Steps:**\n    *   Develop a resource repository of varying complexity levels for each topic.\n    *   Experiment with different media formats (videos, interactive simulations, audio explanations) to cater to diverse learning preferences.\n    *   Offer a range of choices in how students approach their learning - which project they want to work on, which practice questions they attempt, etc.\n\n**3. Algorithm-Based Feedback:**\n\n*   **Timely and Specific Feedback:** Aim to provide feedback that is not only timely but also very specific. Think of it as debugging a code – the feedback needs to pinpoint the problem precisely.\n*   **Actionable Steps:**\n    *   Use rubrics for grading that clearly outline expectations.\n    *   Provide feedback that is focused on the concepts, not just the final answer.\n    *   Use automated tools, where possible, to offer initial feedback, allowing you to focus on more complex issues.\n\n**4. Collaborative Problem-Solving:**\n\n*   **Agile Team Teaching:** Approach your classroom as an agile team working on a project. Encourage collaboration and peer-to-peer learning.\n*   **Actionable Steps:**\n    *   Implement group activities with defined roles, as you might in a project team.\n    *   Have students teach each other concepts, thereby reinforcing their own understanding.\n    *   Use pair programming or group programming tasks to foster learning from each other.\n\n**5. Iterative Teaching Process:**\n\n*   **Continuous Improvement:** Just like in software development, use a continuous improvement approach to your teaching. Regularly evaluate your teaching practices and make adjustments based on results.\n*   **Actionable Steps:**\n    *   Solicit student feedback on your teaching methods and the content.\n    *   Periodically reflect on what works and what doesn't.\n    *   Don't be afraid to experiment with new approaches based on feedback and data.\n\nAaron, your experience with AI and data analysis can be a powerful asset in personalizing your teaching approach. By treating your classroom as a dynamic system with evolving needs, you can create a learning environment that is truly tailored to each of your students. You have an analytical mindset and the tools to make data driven decisions in the classroom. By embracing these strategies, you'll be able to provide a more engaging and effective learning experience for your students! Let me know if you want to explore any of these strategies further.\n"
            }
         },
         {
            "role":"human",
            "type":"text",
            "timestamp":"string",
            "payload":{
               "text":"Please, summarize what you said and translate that to spanish"
            }
         }
      ]
   }
}
```

Response:
image

## Unit Tests

Added new unit tests to cover:
- CoTeacher collaborative scenarios.
- Integration tests to validate smooth workflow functionality.

## Documentation Updates

Indicate whether documentation needs to be updated due to this PR.

- [x] Yes  
  - Added documentation for the new CoTeacher module.  
  - Revised examples to include updated functionality.  

- [ ] No

## Checklist

- [x] I have performed a self-review of my code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [x] Any dependent changes have been merged and published in downstream modules.

## Additional Information

These updates aim to enhance the assistant's capabilities, improve user experience, and provide robust collaborative features through the new CoTeacher implementation.

### CODE REVIEW

File: app/assistants/classroom_support/co_teacher/metadata.json [hunk 1]
```diff
@@ -1,8 +1,18 @@
...
+            "type": "text"
+        },
+        {
+            "label": "User's age",
+            "name": "user_age",
+            "type": "number"
+        },
+        {
+            "label": "User's preference",
+            "name": "user_preference",
```
Comment (yunusj) [inline app/assistants/classroom_support/co_teacher/metadata.json]: Lets make this global. We dont need different inputs per assistant, they'll all be the same.

File: app/utils/actions_for_assistants/actions_for_assistants.py [hunk 2]
```diff
@@ -21,117 +31,251 @@ def read_text_file(file_path):
...
+
+    chain = prompt | chat_google_genai
+    result = chain.invoke({
+        "user_query": user_query,
+        "focus": focus,
+        "tone": tone
+    })
+    return result.content
+
+def custom_prompt_tool(user_query: str, action: str):
```
Comment (yunusj) [inline app/utils/actions_for_assistants/actions_for_assistants.py]: lets remove this for the MVP

File: app/utils/actions_for_assistants/actions_for_assistants.py [hunk 3]
```diff
@@ -21,117 +31,251 @@ def read_text_file(file_path):
...
+    result = combined_chain.invoke(user_query)
 
-    result = chain.invoke(
-        {
-            "questions": state["result"],
-            "format_instructions": parser.get_format_instructions()
-        }
+    return result
+
+def call_marvel_ai_tool(user_query: str):
```
Comment (yunusj) [inline app/utils/actions_for_assistants/actions_for_assistants.py]: Good POC but we need to investigate this a bit more and come up with a more declarative way where our LLM can make better interpretation of what tool to call, just like it does for other actions like summarize etc. Lets omit this feature in this PR

**Review by yunusj — CHANGES_REQUESTED** [review]
Feedback points
1. Omit actions in the MVP until we have understood the UX better
2. Inputs will be same for all assistants so no need for metadata.json per assistant
3. Remove features folder as we don't need that now after having tools and assistants separately

So essentially we are just keeping 1 assistant, very lightweight, with just system prompt being unique per assistant, no actions or tools calling for now.

File: app/assistants/classroom_support/co_teacher/metadata.json (hunk 1 above)
Comment (AaronSosaRamos) [reply app/assistants/classroom_support/co_teacher/metadata.json]: Done

File: app/utils/actions_for_assistants/actions_for_assistants.py (hunk 2 above)
Comment (AaronSosaRamos) [reply app/utils/actions_for_assistants/actions_for_assistants.py]: Done

File: app/utils/actions_for_assistants/actions_for_assistants.py (hunk 3 above)
Comment (AaronSosaRamos) [reply app/utils/actions_for_assistants/actions_for_assistants.py]: Done

**Review by yunusj — APPROVED** [review]
Looks good to go  🚀
Amazing work, @AaronSosaRamos :)

**Pull request merged**